core/services/chainlink/database_config.go

```diff
@@ -0,0 +1,114 @@
+package chainlink
+
+import (
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/chainlink/v2/core/config"
+	v2 "github.com/smartcontractkit/chainlink/v2/core/config/v2"
+	"github.com/smartcontractkit/chainlink/v2/core/store/dialects"
+)
+
+type backupConfig struct {
+	c v2.DatabaseBackup
+	s v2.DatabaseSecrets
+}
+
+func (b *backupConfig) Dir() string {
+	return *b.c.Dir
+}
+
+func (b *backupConfig) Frequency() time.Duration {
+	return b.c.Frequency.Duration()
+}
+
+func (b *backupConfig) Mode() config.DatabaseBackupMode {
+	return *b.c.Mode
+}
+
+func (b *backupConfig) OnVersionUpgrade() bool {
+	return *b.c.OnVersionUpgrade
+}
+
+func (b *backupConfig) URL() *url.URL {
+	return b.s.BackupURL.URL()
+}
+
+var _ config.Database = (*databaseConfig)(nil)
+
+type databaseConfig struct {
+	c  v2.Database
+	s  v2.DatabaseSecrets
+	mu *sync.RWMutex
+}
+
+func (d *databaseConfig) Backup() config.Backup {
+	return &backupConfig{
+		c: d.c.Backup,
+		s: d.s,
+	}
+}
+
+func (d *databaseConfig) DatabaseDefaultIdleInTxSessionTimeout() time.Duration {
+	return d.c.DefaultIdleInTxSessionTimeout.Duration()
+}
+
+func (d *databaseConfig) DatabaseDefaultLockTimeout() time.Duration {
+	return d.c.DefaultLockTimeout.Duration()
+}
+
+func (d *databaseConfig) DatabaseDefaultQueryTimeout() time.Duration {
+	return d.c.DefaultQueryTimeout.Duration()
+}
+
+func (d *databaseConfig) DatabaseListenerMaxReconnectDuration() time.Duration {
+	return d.c.Listener.MaxReconnectDuration.Duration()
+}
+
+func (d *databaseConfig) DatabaseListenerMinReconnectInterval() time.Duration {
+	return d.c.Listener.MinReconnectInterval.Duration()
+}
+
+func (d *databaseConfig) DatabaseLockingMode() string {
+	return d.c.LockingMode()
+}
+
+func (d *databaseConfig) DatabaseURL() url.URL {
+	return *d.s.URL.URL()
+}
+
+func (d *databaseConfig) GetDatabaseDialectConfiguredOrDefault() dialects.DialectName {
+	return d.c.Dialect
+}
+
+func (d *databaseConfig) LeaseLockDuration() time.Duration {
+	return d.c.Lock.LeaseDuration.Duration()
+}
+
+func (d *databaseConfig) LeaseLockRefreshInterval() time.Duration {
+	return d.c.Lock.LeaseRefreshInterval.Duration()
+}
+
+func (d *databaseConfig) MigrateDatabase() bool {
+	return *d.c.MigrateOnStartup
+}
+
+func (d *databaseConfig) ORMMaxIdleConns() int {
+	return int(*d.c.MaxIdleConns)
+}
+
+func (d *databaseConfig) ORMMaxOpenConns() int {
+	return int(*d.c.MaxOpenConns)
+}
+
+func (d *databaseConfig) TriggerFallbackDBPollInterval() time.Duration {
+	return d.c.Listener.FallbackPollInterval.Duration()
+}
+
+func (d *databaseConfig) LogSQL() (sql bool) {
+	d.mu.RLock()
+	sql = *d.c.LogQueries
```

Comment:
Are we still referencing the original field at this point? I would have expected the embedded `v2.Database` to need to be a pointer in that case.

Reply:
And we can leverage that type's method, without adding any extra locking here at all, IIUC

Reply:
I'm fairly certain we are, see this minimal reproduction:

https://go.dev/play/p/jiiMmDpMvE9

I don't think this is unit tested so let me add a test here to be 100%. This also isn't the most obvious way of writing this, so let me think about ways we could make it clearer.

Unfortunately leveraging that type's method isn't an option here since that pointer is referring back to the TOML struct. There might be a better

Reply:
(I refactored this to borrow the LogSQL method from generalConfig, and added a test. The actual behaviour of the mutex is really difficult to test for obvious reasons, but it's something at least.